File: backend/src/main/java/com/draazy/api/leads/notes/LeadNoteService.java
package com.draazy.api.leads.notes;

import com.draazy.api.common.error.BadRequestException;
import com.draazy.api.common.persistence.ConstraintViolations;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.atomic.LongAdder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.PlatformTransactionManager;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.transaction.support.TransactionSynchronizationManager;
import org.springframework.transaction.support.TransactionTemplate;

/**
 * The owner's private lead desk: a note and a follow-up date against any row in their Requests inbox.
 *
 * <p><strong>No mapper, no cross-context lookups, no gate.</strong> Every other service in
 * {@code leads} joins a listing and a user to project a row; this one does not, because the
 * annotation is the owner's own words about a lead they can already see. The gate that matters is
 * the one already applied: {@code ownerId} comes from the JWT and leads every query, so there is no
 * reachable path to another owner's note — not a 403, simply nothing to address.
 */
@Service
public class LeadNoteService {

    private static final Logger log = LoggerFactory.getLogger(LeadNoteService.class);

    /** Exactly as V119 declares it — {@link ConstraintViolations} matches on the name. */
    private static final String UQ_OWNER_LEAD = "uq_lead_notes_owner_lead";

    /**
     * How many leads one owner may annotate.
     *
     * <p><strong>This ceiling is what keeps {@link LeadNoteRepository#findByOwnerId} safe to leave
     * unpaged.</strong> That method's premise is that the collection is bounded by the owner's own
     * typing — true of a person, false of a script. A loop minting a fresh key each time inserts a
     * new row every time, and the read that then has to materialise them all is a {@code GET}, which
     * {@code WriteRateLimitFilter} does not count. So the writes are throttled and the read is not:
     * spend a night growing the table, then replay the cheap request whenever you want the instance
     * to run out of heap. Capping the rows removes the ammunition rather than rationing the trigger.
     *
     * <p>Two thousand is far above any real desk — an owner with two thousand distinct leads they
     * have personally written notes against is not a user this product has — and far below the size
     * where the unpaged read is expensive. Only the <em>insert</em> branch consults it, so an owner
     * who somehow reaches the ceiling can still edit and clear everything they already wrote; the
     * cap can never strand them with a desk they cannot tidy.
     */
    static final int MAX_NOTES_PER_OWNER = 2_000;

    /**
     * How far out of the present a follow-up may sit, in either direction.
     *
     * <p>Not a product opinion about diary length — it is the storage talking. {@link Instant}
     * accepts years up to ±1,000,000,000; {@code timestamptz} runs from 4713 BC to 294276 AD.
     * Without a bound, a body carrying {@code "+1000000000-12-31T23:59:59Z"} reaches the driver and
     * fails there, so any token holder can turn a validation problem into a 500. <strong>The bound
     * is symmetric because the hole is</strong>: {@code "-1000000000-01-01T00:00:00Z"} is the same
     * defect arriving from the other side, and a guard that only looked forward would leave it open
     * while reading as though it had closed it.
     *
     * <p>A past follow-up date is perfectly legitimate — it is an overdue one, which is exactly what
     * the panel's chip is for — so this must not be confused with {@code @FutureOrPresent}. A
     * century in each direction is comfortably inside what Postgres can hold and comfortably outside
     * anything anyone will schedule or backdate.
     *
     * <p>Expressed in days because {@link ChronoUnit#YEARS} is not a unit {@link Instant} supports —
     * years are not a fixed length, and an {@code Instant} has no calendar to resolve them against.
     */
    private static final long MAX_FOLLOW_UP_DAYS = 36_500;

    private final LeadNoteRepository notes;
    private final TransactionTemplate transactions;
    private final LongAdder racesRetried = new LongAdder();

    public LeadNoteService(LeadNoteRepository notes, PlatformTransactionManager transactionManager) {
        this.notes = notes;
        this.transactions = new TransactionTemplate(transactionManager);
    }

    /**
     * How many times the retry below has actually run.
     *
     * <p>Exists because the race test cannot otherwise tell a retry from an accidental
     * serialisation — if the winner commits before the loser's probe, the loser takes the update
     * branch and the observable result is identical to a successful retry. A test that cannot tell
     * those apart silently stops testing this method the first time the timing shifts.
     */
    long racesRetried() {
        return racesRetried.sum();
    }

    /** Every annotation the owner has written, for the inbox to index by key. */
    @Transactional(readOnly = true)
    public List<LeadNoteResponse> myNotes(UUID ownerId) {
        return notes.findByOwnerId(ownerId).stream().map(LeadNoteResponse::of).toList();
    }

    /**
     * Write one annotation, or clear it.
     *
     * <p><strong>The loser of an upsert race is retried, not refused</strong> — the same shape as
     * {@code ConversationOpeningService#start} (D54), and for a sharper reason. There, two people
     * had to message each other at once; here it takes one person, because the panel saves the note
     * and the follow-up date from separate controls. A double-tapped Save, or a blur-autosave on
     * each field, sends two writes on the same key milliseconds apart: both probes miss, both
     * insert, and {@code uq_lead_notes_owner_lead} rejects the second. Left alone that surfaces as a
     * 409 "conflicts with existing data" — a message about contention, for an annotation nobody else
     * can touch, with the owner's typed note lost behind it.
     *
     * <p>The retry cannot be a catch block inside a transaction. The violation is raised at flush,
     * which poisons the persistence context and marks the transaction rollback-only, so a re-read
     * there would fail again at commit and turn the 409 into a 500. The first attempt has to
     * <em>end</em> before the winner's committed row is visible, which is why this method is not
     * {@code @Transactional} and each attempt runs in its own {@link TransactionTemplate}. The
     * {@code isActualTransactionActive} guard keeps that reasoning true if a transactional caller is
     * ever added: the template would join their transaction rather than open one, and the retry
     * would re-enter the very context that cannot be reused — so it rethrows, and the caller gets
     * the honest 409 instead of a 500 nobody would trace back here.
     *
     * @return the stored annotation, or {@link Optional#empty()} when the write cleared it
     */
    public Optional<LeadNoteResponse> save(
            UUID ownerId, String leadKey, String note, Instant followUpAt) {
        if (followUpAt != null
                && (followUpAt.isAfter(Instant.now().plus(MAX_FOLLOW_UP_DAYS, ChronoUnit.DAYS))
                        || followUpAt.isBefore(
                                Instant.now().minus(MAX_FOLLOW_UP_DAYS, ChronoUnit.DAYS)))) {
            throw new BadRequestException(
                    "followUpAt is too far from today to be a date anyone meant to pick.");
        }
        try {
            return transactions.execute(tx -> upsert(ownerId, leadKey, note, followUpAt));
        } catch (DataIntegrityViolationException raced) {
            if (!ConstraintViolations.isOn(raced, UQ_OWNER_LEAD)
                    || TransactionSynchronizationManager.isActualTransactionActive()) {
                throw raced;
            }
            log.info("Concurrent lead-note writes raced on one key for owner {}; retrying as an"
                    + " update against the row the winner inserted", ownerId);
            racesRetried.increment();
            return transactions.execute(tx -> upsert(ownerId, leadKey, note, followUpAt));
        }
    }

    /**
     * <strong>An empty annotation is a delete, not a blank row.</strong> The localStorage version
     * pruned for tidiness; here it is load-bearing twice over. The V119 CHECK rejects a row with
     * neither field, so storing one would be a 500 rather than a no-op; and the inbox answers "does
     * this lead have a note" by existence, so a blank row would light a follow-up chip that points at
     * nothing.
     *
     * <p>Blank strings are folded to {@code null} on the way in. A {@code note} of {@code ""} and a
     * {@code note} of {@code null} mean the same thing to every reader, and letting both into the
     * column would mean the CHECK — and the delete rule above — disagreed with the UI about which
     * annotations are empty.
     *
     * <p>{@code saveAndFlush} rather than {@code save}, because on the update branch the row is
     * already managed: {@code save} merges without flushing, and {@code @UpdateTimestamp} fires at
     * flush. The response would carry the timestamp of the <em>previous</em> edit — which is the one
     * field here whose entire job is to say when this edit happened.
     */
    private Optional<LeadNoteResponse> upsert(
            UUID ownerId, String leadKey, String note, Instant followUpAt) {
        String trimmed = (note == null || note.isBlank()) ? null : note.trim();

        Optional<LeadNote> existing = notes.findByOwnerIdAndLeadKey(ownerId, leadKey);
        if (trimmed == null && followUpAt == null) {
            existing.ifPresent(notes::delete);
            return Optional.empty();
        }

        if (existing.isEmpty() && notes.countByOwnerId(ownerId) >= MAX_NOTES_PER_OWNER) {
            throw new BadRequestException(
                    "You have notes against " + MAX_NOTES_PER_OWNER + " leads already. Clear one you"
                            + " no longer need before adding another.");
        }

        LeadNote row = existing.orElseGet(() -> new LeadNote(ownerId, leadKey));
        row.replace(trimmed, followUpAt);
        return Optional.of(LeadNoteResponse.of(notes.saveAndFlush(row)));
    }
}

File: frontend/src/services/leadNoteService.js
/**
 * Lead Note Service — the owner's private annotations on their inbox.
 *
 * ## What a lead note is
 *
 * The Requests inbox is a lightweight lead desk. Against any row — a number request, a photo
 * request, a document request, a flatmate enquiry — the owner can jot a private note and set a
 * follow-up date. Nobody else ever sees either field: not the buyer the note is about, not staff,
 * not admin. That privacy is the whole feature, and it is why the store is caller-scoped rather
 * than property-scoped.
 *
 * ## Keyed by lead, not by row id
 *
 * A note hangs off a `leadKey` the client mints — `'number:<contactRequestId>'`,
 * `'photo:<photoRequestId>'`, `'documents:<buyerMobile>|<propertyId>'`, `'flatmate:<requestId>'` —
 * because a "lead" is not one row. The same buyer asking for the number, then photos, then papers
 * is one conversation the owner is tracking, and the document lead in particular has no single row
 * to point at. The server never parses a key; it stores the string and enforces uniqueness per
 * caller. That keeps the vocabulary of what counts as a lead entirely on this side, which is right,
 * because it is a UI concern.
 *
 * ## Shape
 *
 * Both providers return the server's row, so no call site can tell which one it is talking to:
 *
 *   { leadKey, note, followUpAt, updatedAt }
 *
 *   note        null when the owner set only a follow-up date
 *   followUpAt  null when they wrote only a note; never both null (the row would not exist)
 *   updatedAt   what the localStorage version also called `updatedAt`; the panel shows it
 *
 * ## Whole annotation in, not a patch
 *
 * `save` takes the complete annotation, but the two controls in the sheet edit one field each. The
 * merge therefore happens at the call site, which holds the current value — see `EnquiriesPanel`.
 * That is not a convenience: JSON cannot distinguish "field omitted" from "field cleared to null",
 * so a partial write has no way to express *clearing* the follow-up date, and every encoding that
 * fixes it costs more than merging one object.
 *
 * ## Clearing is a write
 *
 * An annotation with neither field is not stored as a blank row — the row is deleted, and `save`
 * resolves to `null`. There is deliberately no `remove`: the owner clears a note by emptying it,
 * which is the same gesture as editing it, and a second verb would be a second way to say one
 * thing.
 *
 * This replaced `lib/leadNotes.js`, which kept annotations in localStorage under
 * `draazyLeadNotes:<ownerDigits>`. That store was per-browser, so an owner who answered enquiries
 * on their phone and their laptop kept two disjoint sets of notes and neither was wrong — which is
 * the worst shape a CRM field can have.
 */
import { createProvider } from './config.js';

const provider = createProvider('leadNote');

/**
 * Every annotation the caller has written, as an array.
 *
 * Unpaged, and unpaged on purpose: the inbox needs the whole set at once because it indexes them by
 * `leadKey` to decorate rows, so a page boundary would silently blank the notes on some rows and
 * not others. The ceiling that makes this safe is enforced server-side on write.
 *
 * @returns {Promise<Array<{leadKey: string, note: ?string, followUpAt: ?string, updatedAt: string}>>}
 */
export const myLeadNotes = async () => (await provider()).myLeadNotes();

/**
 * Write one annotation, or clear it.
 *
 * @param {string} leadKey the client-minted lead identity; see the module note
 * @param {{note?: ?string, followUpAt?: ?string}} annotation the *whole* annotation, not a patch
 * @returns {Promise<?object>} the stored row, or `null` when the annotation was empty and the row
 *   was therefore deleted. Callers must handle `null` — it is the normal result of clearing a note,
 *   not a failure.
 */
export const saveLeadNote = async (leadKey, annotation) => (await provider()).saveLeadNote(leadKey, annotation);
